File: api/permissions.py
"""Permission classes for the mobile API.

The web app authorizes by *page*: ``WebAccessMiddleware`` maps a URL name to a
tab right. This module used to say that model does not apply to resource APIs,
and left ``IsAuthenticated`` as the whole policy. It does apply, and leaving it
out meant any valid token could create, edit and delete across 54 writable
resources regardless of the group matrix governing the same data in the web app
— the API was a way round the entire permission system.

``MatrixPermission`` closes that by mapping a resource to the tab that owns it
and the HTTP method to an action the matrix already understands. Data scoping
(branch / farm / warehouse ownership) is applied separately, on the queryset —
see ``api.viewsets``.

``ReadOnlyOrStaff`` remains for master/reference data that mobile should read
but not mutate.
"""
from __future__ import annotations

import logging

from django.conf import settings
from rest_framework.permissions import BasePermission, IsAuthenticated, SAFE_METHODS

__all__ = ["IsAuthenticated", "ReadOnlyOrStaff", "MatrixPermission", "MODEL_TABS"]

logger = logging.getLogger(__name__)


class ReadOnlyOrStaff(BasePermission):
    """Anyone authenticated may read; only staff may write."""

    def has_permission(self, request, view) -> bool:
        if not (request.user and request.user.is_authenticated):
            return False
        if request.method in SAFE_METHODS:
            return True
        return bool(request.user.is_staff or request.user.is_superuser)


METHOD_ACTIONS = {
    "GET": "view", "HEAD": "view", "OPTIONS": "view",
    "POST": "add", "PUT": "edit", "PATCH": "edit", "DELETE": "delete",
}

#: "app_label.ModelName" -> the tab code that owns it.
#:
#: Only what is genuinely one-to-one. A wrong entry refuses a request the web
#: app would allow, which is worse than an unmapped one, so anything ambiguous
#: is left out and turns up in the audit instead.
MODEL_TABS = {
    # Broiler
    "broiler.Region": "region",
    "broiler.Branch": "branch_template",
    "broiler.Supervisor": "supervisor_template",
    "broiler.BroilerLine": "broiler_line",
    "broiler.BroilerFarm": "branch_farm",
    "broiler.BroilerFarmShed": "broiler_farm_shed",
    "broiler.BroilerBatch": "broiler_batch",
    "broiler.FarmerGroup": "farmer_group",
    "broiler.Farmer": "branch_farm",
    "broiler.DailyEntry": "daily_entry_list",
    "broiler.BirdSale": "bird_sale_list",
    "broiler.BirdSaleReceipt": "bird_sale_receipt_list",
    "broiler.BroilerDisease": "broiler_disease",
    "broiler.Breed": "breed",
    "broiler.BreedStandard": "breed_standard",
    "broiler.FarmLocationCapture": "farm_location_capture_list",
    "broiler.MedicineVaccineEntry": "medicine_entry_list",
    "broiler.GrowingChargeScheme": "growing_charge",
    "broiler.GrowingChargeSettlement": "gc_settlement",
    "broiler.FarmerFarmSetupRequest": "farmer_farm_setup_request_list",
    # Photos belong to the entry they are attached to — same tab, so a user who
    # may see the record may see its pictures and no one else can.
    "broiler.BirdSalePhoto": "bird_sale_list",
    "broiler.DailyEntryPhoto": "daily_entry_list",
    # Inventory
    "inventory.Item": "items",
    "inventory.ItemCategory": "item_category",
    "inventory.Warehouse": "warehouse",
    "inventory.StockTransfer": "stock_transfer_list",
    "inventory.StockIssue": "stock_issue_list",
    "inventory.StockReceive": "stock_receive_list",
    "inventory.InventoryAdjustment": "inventory_adjustment_list",
    "inventory.MedicineTransfer": "medicine_transfer_list",
    "inventory.Sector": "sector",
    "inventory.UnitOfMeasurement": "unit_of_measurement",
    "inventory.ItemPriceList": "item_price_list",
    # Purchase
    "purchase.Supplier": "supplier",
    "purchase.VendorGroup": "vendor_groups",
    "purchase.GeneralPurchase": "general_purchase_list",
    "purchase.ChicksPurchase": "chicks_purchase_list",
    "purchase.DebitNote": "debit_note_list",
    "purchase.CreditNote": "credit_note_list",
    "purchase.SupplierPayment": "payment_list",
    "purchase.TaxMaster": "tax_master",
    # Line items carry the tab of the document they belong to.
    "purchase.GeneralPurchaseItem": "general_purchase_list",
    "purchase.ChicksPurchaseItem": "chicks_purchase_list",
    "purchase.SupplierPaymentLine": "payment_list",
    # Sales
    "sales.Customer": "customer",
    "sales.CustomerGroup": "customer_groups",
    "sales.SalesInvoice": "sales_invoice_list",
    "sales.SalesInvoiceItem": "sales_invoice_list",
    "sales.SalesReceipt": "sales_receipt_list",
    "sales.SalesPriceMaster": "sales_price_master",
    # Hatchery
    "hatchery.EggPurchase": "egg_purchase_list",
    "hatchery.ChickSale": "chick_sale_list",
    "hatchery.DeliveryChallan": "delivery_challan_list",
    "hatchery.EggGrading": "egg_grading_list",
    "hatchery.HatchSetting": "hatchery_list",
    "hatchery.TraySetting": "tray_set_list",
    "hatchery.HatchEntry": "hatch_entry_list",
    "hatchery.ChangeRequest": "change_requests",
    "hatchery_master.ExpenseType": "expense_type_list",
    "hatchery_master.HatcheryExpense": "hatchery_expense_list",
    "hatchery.EggPurchaseItem": "egg_purchase_list",
    "hatchery.ChickSaleItem": "chick_sale_list",
    "hatchery.DeliveryChallanItem": "delivery_challan_list",
    # Account
    "account.ChartOfAccount": "coa",
    "account.Voucher": "vouchers",
    "account.FinancialYear": "fin_year",
    "account.BankCashMaster": "bank_cash",
    "account.OrganizationCentre": "organization_centre",
    "account.CompanyProfile": "company_profile",
    "account.TermsConditions": "terms",
    # HR
    "hr.Employee": "employee_list",
    "hr.Designation": "designation",
    "hr.Attendance": "employee_attendance",
    "hr.EmployeeLeave": "leave_employee",
    "hr.LeaveSelectedDate": "employee_leave_details",
    "hr.Payroll": "payroll",
    "hr.Group": "employee_group",
    # SMS
    "notification.SmsTemplate": "sms_templates",
    "notification.SmsMessage": "sms_history",
    "notification.SmsSettings": "sms_settings",
    # Hatchery equipment. MASTER_REFERENCE_TABS already names these three, but
    # a tab there does nothing until the model resolves to it — unmapped, they
    # were reaching the picker through the unmapped-is-allowed fallback, which
    # the WEB_ACCESS_ENFORCE rollout removes. The Hatch Setting form's hatchery
    # picker would have gone blank on the day that flag was turned on.
    "hatchery_master.Hatchery": "hatchery_master_list",
    "hatchery_master.Hatcher": "hatcher_list",
    "hatchery_master.Setter": "setter_list",
}


#: Resources with no tab because no web page owns them.
#:
#: These pass today only because an unmapped resource is allowed, which is the
#: audit's fallback and goes away the day ``WEB_ACCESS_ENFORCE`` is turned on —
#: at which point every one of these screens would start refusing everybody.
#: Naming them converts an accidental pass into a decided one, so the flag can
#: be flipped without taking them down.
#:
#: The reasons are ``mobile_access.UNGATED_SCREENS``', already reviewed there:
#: no web page exists, so there is no permission to inherit, and a driver's own
#: vehicles are their reference data — the API narrows each of these to the
#: caller rather than leaving them wide.
UNGATED_MODELS = {
    "hr.Department",
    "hr.Shift",
    "hr.EmployeeVehicle",
    # Trips and their visits are narrowed to the caller by the viewset
    # itself ("the list shows their trips and no one else's"), so the
    # Home screen's own-round widget must not need the trip *report*
    # right to ask. Mapping them to that tab took a supervisor's own
    # day away from them, which is how this was found.
    "hr.SupervisorTrip",
    "hr.SupervisorTripVisit",
    "purchase.CreditTerm",
    "purchase.SupplierShippingAddress",
    "sales.CustomerShippingAddress",
}


def model_for_view(view):
    model = getattr(getattr(view, "queryset", None), "model", None)
    if model is None:
        serializer = getattr(view, "serializer_class", None)
        model = getattr(getattr(serializer, "Meta", None), "model", None)
    return model


def tab_for_view(view):
    """The tab a viewset belongs to, or None when it is not mapped yet.

    An explicit ``tab_code`` on the viewset wins, so a resource can opt in
    without waiting for the map.
    """
    explicit = getattr(view, "tab_code", None)
    if explicit:
        return explicit
    model = model_for_view(view)
    if model is None:
        return None
    return MODEL_TABS.get(f"{model._meta.app_label}.{model.__name__}")


class MatrixPermission(BasePermission):
    """Hold an API request to the same matrix the web app uses.

    Resources with no tab mapped yet are **recorded, not refused**, exactly as
    the middleware handles the unmapped URL surface: filling the map in blind
    would break the mobile client for endpoints nobody has claimed. Set
    ``WEB_ACCESS_ENFORCE=True`` once ``manage.py webaccess_audit`` is clean.
    """

    message = "You do not have permission to use this resource."

    def has_permission(self, request, view) -> bool:
        from user.access import user_can

        user = getattr(request, "user", None)
        if not user or not user.is_authenticated:
            return False

        tab = tab_for_view(view)
        if tab is None:
            return self._unmapped(request, view, user)

        action = METHOD_ACTIONS.get(request.method, "view")

        # Reading a master is part of using the transaction that asks for it.
        # The phone's Daily Entry screen needs farms, batches and supervisors
        # to fill its pickers; refusing those because the user was not also
        # given the Broiler Farm master left the screen open with every
        # dropdown empty and nothing to say why. Same rule the web middleware
        # applies to the master JSON feeds — read is opened, writing a master
        # still needs that master's own rights, and the rows that come back are
        # still narrowed by the data scope.
        from user.access import MASTER_REFERENCE_TABS

        if action == "view" and tab in MASTER_REFERENCE_TABS:
            return True

        # Parties and ledgers are not in that set and must not be — they carry
        # contact, credit and payroll columns, and it opens a read to everyone
        # with a login. They get the narrower rule instead: the transaction the
        # picker belongs to is what entitles the read, exactly as it does on the
        # web, where the form is handed its customers through view context and
        # never asks the master. `picker_only` narrows the row to identity.
        from user.access import picker_read_allowed
        from user.services.mobile_access import mobile_can

        if (action == "view" and not user_can(user, tab, action)
                and picker_read_allowed(
                    user, tab, entitles=lambda code: mobile_can(user, code, "view"))):
            request.picker_only = True
            return True

        if not user_can(user, tab, action):
            self._record(request, view, user, "denied", tab, action)
            return False

        # Second gate, phone only: Mobile Access narrows what the matrix allows
        # — a whole module, or one action on one screen. Subtractive, so it is
        # only ever consulted after the matrix has already said yes.
        from user.services.mobile_access import mobile_can

        if not mobile_can(user, tab, action):
            self._record(request, view, user, "denied", tab, action)
            return False
        return True

    def _unmapped(self, request, view, user) -> bool:
        model = model_for_view(view)
        if model is not None and (
                f"{model._meta.app_label}.{model.__name__}" in UNGATED_MODELS):
            return True             # decided, not merely unmapped

        self._record(request, view, user, "unmapped", "", "")
        if not getattr(settings, "WEB_ACCESS_ENFORCE", False):
            return True             # audit only — the API behaves as before
        logger.warning("api: refusing unmapped resource %s for %s",
                       view.__class__.__name__, user.get_username())
        return False

    def _record(self, request, view, user, verdict, tab, action) -> None:
        from django.core.cache import cache
        from django.db.models import F

        from user.models import WebAccessAudit

        name = f"api:{view.__class__.__name__}"
        username = user.get_username()
        throttle = f"wa:audit:{verdict}:{name}:{request.method}:{username}"
        if cache.get(throttle):
            return
        cache.set(throttle, 1, 600)
        try:
            updated = WebAccessAudit.objects.filter(
                url_name=name, method=request.method, verdict=verdict,
                username=username).update(hits=F("hits") + 1)
            if not updated:
                WebAccessAudit.objects.create(
                    url_name=name, method=request.method, verdict=verdict,
                    username=username, path=request.path[:300],
                    view=f"{view.__class__.__module__}.{view.__class__.__name__}"[:200],
                    tab_code=tab, action=action)
        except Exception:
            # Auditing must never be why an API call fails.
            logger.exception("api: could not record %s for %s", verdict, name)

File: mobile/src/api/permissions.ts
import { http } from "./client";
import { Envelope } from "./types";

export interface ModuleActions {
  add: boolean;
  edit: boolean;
  delete: boolean;
}

/** Per-screen actions: what may be done inside one hub tile. */
export interface TabActions {
  view: boolean;
  add: boolean;
  edit: boolean;
  delete: boolean;
}

export interface Permissions {
  unrestricted: boolean;
  nav_groups: string[];
  /** Module keys in home-hub order — administrator order where one is set,
   *  registry order otherwise. Never alphabetical. */
  nav_order: string[];
  tabs: string[];
  module_actions: Record<string, ModuleActions>;
  /** tab_code -> the four actions, already narrowed by Mobile Access. */
  tab_actions: Record<string, TabActions>;
}

export type ActionKind = keyof ModuleActions;

export async function fetchPermissions(): Promise<Permissions> {
  const resp = await http.get<Envelope<Permissions>>("/auth/permissions");
  return resp.data.data;
}

/**
 * Mobile resource key → backend tab_code, for finer per-tab view gating of hub
 * tiles. Resources without an entry fall back to module-level access (shown
 * whenever the module itself is accessible). Line-item children aren't listed
 * (they render inside a parent, not the hub).
 */
export const RESOURCE_TABS: Record<string, string> = {
  // Broiler
  "broiler-daily-entries": "daily_entry_list",
  "broiler-medicine-vaccine": "medicine_entry_list",
  "broiler-bird-sales": "bird_sale_list",
  "broiler-sale-receipts": "bird_sale_receipt_list",
  "broiler-farms": "branch_farm",
  "broiler-sheds": "broiler_farm_shed",
  "broiler-batches": "broiler_batch",
  "broiler-farmer-groups": "farmer_group",
  "broiler-regions": "region",
  "broiler-branches": "branch_template",
  "broiler-lines": "broiler_line",
  "broiler-supervisors": "supervisor_template",
  "broiler-breeds": "breed",
  "broiler-breed-standards": "breed_standard",
  "broiler-diseases": "broiler_disease",
  "broiler-growing-charges": "growing_charge",
  "broiler-gc-settlements": "gc_settlement",
  // Hatchery
  "hatchery-egg-purchases": "egg_purchase_list",
  "hatchery-egg-gradings": "egg_grading_list",
  "hatchery-hatch-settings": "hatchery_list",
  "hatchery-tray-settings": "tray_set_list",
  "hatchery-hatch-entries": "hatch_entry_list",
  "hatchery-chick-sales": "chick_sale_list",
  "hatchery-delivery-challans": "delivery_challan_list",
  "hatchery-expenses": "hatchery_expense_list",
  "hatchery-expense-types": "expense_type_list",
  "hatchery-hatcheries": "hatchery_master_list",
  "hatchery-setters": "setter_list",
  "hatchery-hatchers": "hatcher_list",
  // Inventory
  "inventory-item-categories": "item_category",
  "inventory-items": "items",
  "inventory-price-list": "item_price_list",
  "inventory-sectors": "sector",
  "inventory-uom": "unit_of_measurement",
  "inventory-warehouses": "warehouse",
  "inventory-stock-transfers": "stock_transfer_list",
  "inventory-medicine-transfers": "medicine_transfer_list",
  "inventory-adjustments": "inventory_adjustment_list",
  "inventory-stock-issues": "stock_issue_list",
  "inventory-stock-receives": "stock_receive_list",
  // Account
  "account-financial-years": "fin_year",
  "account-chart-of-accounts": "coa",
  "account-bank-cash": "bank_cash",
  "account-organization-centres": "organization_centre",
  "account-company-profiles": "company_profile",
  "account-terms": "terms",
  "account-vouchers": "vouchers",
  // Sales
  "sales-invoices": "sales_invoice_list",
  "sales-receipts": "sales_receipt_list",
  "sales-customers": "customer",
  "sales-customer-groups": "customer_groups",
  "sales-prices": "sales_price_master",
  // Purchase
  "purchase-general-purchases": "general_purchase_list",
  "purchase-chicks-purchases": "chicks_purchase_list",
  "purchase-supplier-payments": "payment_list",
  "purchase-debit-notes": "debit_note_list",
  "purchase-credit-notes": "credit_note_list",
  "purchase-suppliers": "supplier",
  "purchase-vendor-groups": "vendor_groups",
  "purchase-tax-masters": "tax_master",
  // HR
  "hr-employees": "employee_list",
  "hr-attendance": "employee_attendance",
  "hr-leaves": "leave_employee",
  "hr-leave-dates": "employee_leave_details",
  "hr-payroll": "payroll",
  "hr-designations": "designation",
  "hr-groups": "employee_group",
  // Users
  "user-users": "create_user",
  "user-groups": "user_groups",
  "user-group-permissions": "assign_groups",
  // SMS
  "sms-templates": "sms_templates",
  "sms-messages": "sms_history",
  "sms-settings": "sms_settings",
};

/**
 * Report tile key → backend tab code.
 *
 * The hub's report tiles used to render unconditionally: every user saw every
 * report their module allowed, whatever the web matrix said about the report
 * itself. These are view-only, so they are gated on `canTab` rather than given
 * rows in the Mobile Access matrix — a report has no Add/Edit/Delete to tick,
 * and columns that control nothing are the thing that matrix exists to avoid.
 *
 * `mortality-trend` has no web report behind it, so it stays ungated and is
 * deliberately absent rather than pointed at an approximate tab.
 */
export const REPORT_TABS: Record<string, string> = {
  "live-flock": "live_flock_summary_report",
  "batch-summary": "broiler_batch_report",
  "chicks-placement": "chicks_placement_report",
  "day-record": "day_record_report",
  "feed-dispatch": "feed_dispatch_stock_report",
  "lifting": "lifting_report",
  "hatch-performance": "hatchery_report",
  "egg-intake": "egg_purchase_report",
  "incubation": "incubation_report",
  "delivery-challan": "delivery_challan_report",
  "chick-sale": "chick_sale_report",
};

/** Mobile module key → backend nav group key (SMS lives under "notifications"). */
export const MODULE_NAV: Record<string, string> = {
  broiler: "broiler",
  hatchery: "hatchery",
  sms: "notifications",
  account: "account",
  inventory: "inventory",
  sales: "sales",
  purchase: "purchase",
  hr: "hr",
  user: "user",
};
